### DISCUSSION TITLE
Middleware- Error: The edge runtime does not support Node.js _____  module

### DISCUSSION THREAD

**Discussion by rreeves8** [vercel/next.js#46520] [Help]
Firebase-admin and middleware seem incompatible cause vercel won't support certain node modules ??

I dont understand the purpose of even having a middleware functionality if 90% of npm libraries are riddled unusable. Seems ridiculous I can't verify a simple web token. There's no way I'm the only person irritated by this. It seems vercel has limited the usage maybe for cloud resource savings ? I dont get it.

I think the Next Js team needs to improve their backend functionalities and design patterns. It seems the frameworks mainly focused around Ui and is missing core functionalties or has design flaws for the backend (in my opinion). The fact I have to make a different file for each API route in the new app system is ridiculous, the folder structure becomes unreadable and managing them mentally is a pain. 

I dont understand why they can't make one file called middleware that supports node modules so I can actaully acheive something, and one file called routes that idk imports some object and we can adds routes like in express.

**Comment by icyJoseph** [comment]
This is because people often think of Node.js as the only JavaScript runtime, but of course that's not the case. We have many runtimes.

Runtimes are built on top of engines, for example the V8 engine, is used to build Chromium and Node.js - Vercel uses the V8 engine to build a server side runtime called Edge, which implements standard Web APIs, for example https://developer.mozilla.org/en-US/docs/Web/API/Response, or https://developer.mozilla.org/en-US/docs/Web/API/Request.

This means that often Node.js specific built-ins won't be present, unless implemented by the runtime. 

As far as I know this is also an issue for Deno users? Unfortunately the problem is that Firebase-admin is built to work only with Node.js. 

Perhaps 3 years ago Node.js was the de-facto runtime, but since then several packages and libraries have changed their abstraction to be able to execute in various runtimes. Vercel saw that and decided to surf the wave as well. Why? Because this makes the Middleware code portable and removes the need for cold-starts in a serverless environment. 

I think when emotions get involved it is good to try and stay objective. The Middleware Edge documentation is extensive as to what's supported and what can't really be done. https://nextjs.org/docs/api-reference/edge-runtime

To work with JWTs in Edge, people use `jose`, https://www.npmjs.com/package/jose which is a zero dependency, runtime agnostic library 👀 - how's that not a positive net win for the entire community?

Tech progress is savage, what used to be cutting edge yesterday falls short today. I found a couple of issues in `firebase-admin` where they are sort of waving away Deno, Edge, and not even considering Bun, runtimes... Here's an opportunity that they are just simply choosing to ignore.

- https://github.com/firebase/firebase-admin-node/issues/1307
- https://github.com/firebase/firebase-admin-node/issues/1801
- https://github.com/vuejs/vuefire/issues/1284

Apparently, you could tweak WebPack during `edge` build phase, to make libraries "nodeless", https://github.com/unjs/unenv, but I haven't used it myself.
